File: packages/sdk-python/fossil/store.py
from __future__ import annotations
import json
import sqlite3
from pathlib import Path
from typing import List, Optional, Tuple
import numpy as np

from .schema import FossilRecord
from .embedder import BaseEmbedder, cosine_similarity, get_default_embedder


DEFAULT_DB_PATH = Path.home() / ".fossil" / "fossil.db"


class FossilStore:
    def __init__(
        self,
        db_path: Path | str = DEFAULT_DB_PATH,
        embedder: Optional[BaseEmbedder] = None,
    ):
        self._path = Path(db_path)
        self._path.parent.mkdir(parents=True, exist_ok=True)
        self._embedder = embedder or get_default_embedder()
        self._conn = sqlite3.connect(str(self._path), check_same_thread=False)
        self._init_schema()

    def _init_schema(self) -> None:
        self._conn.executescript("""
            CREATE TABLE IF NOT EXISTS fossils (
                id TEXT PRIMARY KEY,
                protocol_version TEXT NOT NULL,
                timestamp TEXT NOT NULL,
                data JSON NOT NULL,
                embedding BLOB NOT NULL,
                shared INTEGER NOT NULL DEFAULT 0
            );

            CREATE INDEX IF NOT EXISTS idx_shared ON fossils(shared);
            CREATE INDEX IF NOT EXISTS idx_timestamp ON fossils(timestamp);
        """)
        self._conn.commit()

    def insert(self, record: FossilRecord) -> FossilRecord:
        text = self._situation_text(record)
        embedding = self._embedder.embed(text)
        blob = np.array(embedding, dtype=np.float32).tobytes()

        self._conn.execute(
            """
            INSERT OR REPLACE INTO fossils
                (id, protocol_version, timestamp, data, embedding, shared)
            VALUES (?, ?, ?, ?, ?, ?)
            """,
            (
                record.id,
                record.protocol_version,
                record.timestamp,
                json.dumps(record.to_dict()),
                blob,
                int(record.shared),
            ),
        )
        self._conn.commit()
        return record

    def search(
        self,
        situation_text: str,
        top_k: int = 5,
        min_score: float = 0.5,
        domain: Optional[str] = None,
    ) -> List[Tuple[FossilRecord, float]]:
        query_vec = np.array(
            self._embedder.embed(situation_text), dtype=np.float32
        )

        cur = self._conn.execute(
            "SELECT id, data, embedding FROM fossils"
        )
        rows = cur.fetchall()

        scored: List[Tuple[FossilRecord, float]] = []
        for row_id, data_json, blob in rows:
            record = FossilRecord.from_dict(json.loads(data_json))

            if domain and record.agent.task_domain.value != domain:
                continue

            stored_vec = np.frombuffer(blob, dtype=np.float32)
            score = float(np.dot(query_vec, stored_vec))

            if score >= min_score:
                scored.append((record, round(score, 4)))

        scored.sort(key=lambda x: x[1], reverse=True)
        return scored[:top_k]

    def get(self, fossil_id: str) -> Optional[FossilRecord]:
        cur = self._conn.execute(
            "SELECT data FROM fossils WHERE id = ?", (fossil_id,)
        )
        row = cur.fetchone()
        if not row:
            return None
        return FossilRecord.from_dict(json.loads(row[0]))

    def delete(self, fossil_id: str) -> bool:
        cur = self._conn.execute(
            "DELETE FROM fossils WHERE id = ?", (fossil_id,)
        )
        self._conn.commit()
        return cur.rowcount > 0

    def count(self) -> int:
        cur = self._conn.execute("SELECT COUNT(*) FROM fossils")
        return cur.fetchone()[0]

    def list_all(self, limit: int = 100, offset: int = 0) -> List[FossilRecord]:
        cur = self._conn.execute(
            "SELECT data FROM fossils ORDER BY timestamp DESC LIMIT ? OFFSET ?",
            (limit, offset),
        )
        return [FossilRecord.from_dict(json.loads(r[0])) for r in cur.fetchall()]

    def close(self) -> None:
        self._conn.close()

    def _situation_text(self, record: FossilRecord) -> str:
        parts = [record.situation.description, record.failure.description]
        if record.situation.context_snapshot:
            parts.append(record.situation.context_snapshot[:512])
        return " | ".join(parts)

    def __enter__(self) -> FossilStore:
        return self

    def __exit__(self, *_) -> None:
        self.close()

File: apps/api/src/lib/store.ts
import type { FossilRecord } from "./types";
import { cosineSimilarity } from "./embedder";

export interface SearchResult {
  record: FossilRecord;
  score: number;
}

export async function insertFossil(
  db: D1Database,
  record: FossilRecord,
  embedding: number[]
): Promise<void> {
  const embeddingJson = JSON.stringify(embedding);
  await db
    .prepare(
      `INSERT OR REPLACE INTO fossils (
        id, protocol_version, timestamp,
        framework, model, task_domain,
        situation, context_snapshot,
        failure_type, failure_description, severity, was_irreversible,
        resolution_type, resolution_description, verified, time_to_resolve_minutes,
        embedding, shared
      ) VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)`
    )
    .bind(
      record.id,
      record.protocol_version,
      record.timestamp,
      record.agent.framework,
      record.agent.model,
      record.agent.task_domain,
      record.situation.description,
      record.situation.context_snapshot ?? null,
      record.failure.type,
      record.failure.description,
      record.failure.severity,
      record.failure.was_irreversible ? 1 : 0,
      record.resolution.type,
      record.resolution.description,
      record.resolution.verified ? 1 : 0,
      record.resolution.time_to_resolve_minutes ?? null,
      embeddingJson,
      record.shared ? 1 : 0
    )
    .run();
}

export async function searchFossils(
  db: D1Database,
  queryEmbedding: number[],
  topK: number,
  minScore: number,
  domain?: string,
  pool?: string
): Promise<SearchResult[]> {
  const conditions: string[] = [];
  const bindings: unknown[] = [];

  if (domain) {
    conditions.push("task_domain = ?");
    bindings.push(domain);
  }

  if (pool === "community") {
    conditions.push("shared = 1");
  }

  let query = "SELECT * FROM fossils";
  if (conditions.length > 0) {
    query += " WHERE " + conditions.join(" AND ");
  }

  const { results } = await db
    .prepare(query)
    .bind(...bindings)
    .all();

  const scored = results
    .map((row: any) => {
      const embedding = JSON.parse(row.embedding as string) as number[];
      const score = cosineSimilarity(queryEmbedding, embedding);
      return { row, score };
    })
    .filter(({ score }) => score >= minScore)
    .sort((a, b) => b.score - a.score)
    .slice(0, topK);

  return scored.map(({ row, score }) => ({
    record: rowToRecord(row),
    score: Math.round(score * 10000) / 10000,
  }));
}

export async function getFossil(
  db: D1Database,
  id: string
): Promise<FossilRecord | null> {
  const row = await db
    .prepare("SELECT * FROM fossils WHERE id = ?")
    .bind(id)
    .first();
  if (!row) return null;
  return rowToRecord(row as any);
}

export async function deleteFossil(
  db: D1Database,
  id: string
): Promise<boolean> {
  const result = await db
    .prepare("DELETE FROM fossils WHERE id = ?")
    .bind(id)
    .run();
  return (result.meta.changes ?? 0) > 0;
}

export async function listFossils(
  db: D1Database,
  limit: number,
  offset: number
): Promise<FossilRecord[]> {
  const { results } = await db
    .prepare(
      "SELECT * FROM fossils ORDER BY timestamp DESC LIMIT ? OFFSET ?"
    )
    .bind(limit, offset)
    .all();
  return results.map((row) => rowToRecord(row as any));
}

export async function countFossils(db: D1Database): Promise<number> {
  const row = await db
    .prepare("SELECT COUNT(*) as count FROM fossils")
    .first<{ count: number }>();
  return row?.count ?? 0;
}

function rowToRecord(row: any): FossilRecord {
  return {
    id: row.id,
    protocol_version: row.protocol_version,
    timestamp: row.timestamp,
    agent: {
      framework: row.framework,
      model: row.model,
      task_domain: row.task_domain,
    },
    situation: {
      description: row.situation,
      context_snapshot: row.context_snapshot ?? null,
    },
    failure: {
      type: row.failure_type,
      description: row.failure_description,
      severity: row.severity,
      was_irreversible: row.was_irreversible === 1,
    },
    resolution: {
      type: row.resolution_type,
      description: row.resolution_description,
      verified: row.verified === 1,
      time_to_resolve_minutes: row.time_to_resolve_minutes ?? null,
    },
    shared: row.shared === 1,
  };
}